tags_update/verify.py

```
@@ -0,0 +1,67 @@
+SUCCESS_MESSAGE = "Great work! You have successfully updated the tags to point to the correct commits."
+MISSING_COMMIT_MESSAGE = "Could not find a commit with '{message}' in the message"
+
+
+def get_commit_from_message(
+    commits: List[GitAutograderCommit], message: str
+) -> Optional[GitAutograderCommit]:
+    """Find a commit with the given message from a list of commits."""
+    for commit in commits:
+        if message.strip() == commit.commit.message.strip():
+            return commit
+    return None
+
+
+def verify(exercise: GitAutograderExercise) -> GitAutograderOutput:
+    tags = exercise.repo.repo.tags
+
+    if "first-update" in tags:
+        raise exercise.wrong_answer([OLD_FIRST_UPDATE_TAG])
+
+    if "january-update" not in tags:
+        raise exercise.wrong_answer([MISSING_JANUARY_TAG])
+
+    # Ensure january-update tag points to the correct commit
+    main_branch_commits = exercise.repo.branches.branch("main").commits
+    january_commit = get_commit_from_message(
+        main_branch_commits, "Add January duty roster"
+    )
+    if january_commit is None:
+        raise exercise.wrong_answer([MISSING_COMMIT_MESSAGE.format(message="January")])
+
+    january_tag_commit = tags["january-update"].commit
+    if january_tag_commit.hexsha != january_commit.hexsha:
+        raise exercise.wrong_answer([WRONG_JANUARY_TAG_COMMIT])
+
+    if "april-update" not in tags:
+        raise exercise.wrong_answer([MISSING_APRIL_TAG])
+
+    # Ensure april-update tag points to the correct commit
+    april_commit = get_commit_from_message(
+        main_branch_commits, "Update duty roster for April"
+    )
+    if april_commit is None:
+        raise exercise.wrong_answer([MISSING_COMMIT_MESSAGE.format(message="April")])
+
+    april_tag_commit = tags["april-update"].commit
+    if april_tag_commit.hexsha != april_commit.hexsha:
+        raise exercise.wrong_answer([WRONG_APRIL_TAG_COMMIT])
```

Since the user could potentially struggle with multiple issues at once, what do you think of collecting the comments and displaying them all together at the end instead of raising an error at the first mistake? This would require some extra tweaking in the test verification setup. I'd like to hear your thoughts on it.

---

I think it would be good to collate all error messages and return them to the user. However, we should also follow the Task-based Sequential Verification approach (as mentioned above by Jia Hao), where we fully verify Task 1 first before Task 2 (since that is the order that users are likely to follow).

We also should fail early on dependent checks, ie. if this check fails, subsequent checks are likely to fail as well, eg. if cannot find `april_commit`, we can just fail early.

We should only collate comments if they are for multiple independent errors for the same task.

In this case, it seems quite hard to find multiple independent errors for the same task, the only case I can think is for the first task, where I check for both tags `first-update` and `january-update`

---

That's fair.